### PR TITLE
Java: Migrate path injection sinks to models-as-data (simplified)

### PR DESCRIPTION
**Pull request by atorralba**
Migrates path injection sinks to models-as-data format. This is a simplification of #12886 to avoid introducing changes in alerts.

The existing `PathCreation` classes can't be removed because doing so would cause alert wobble in the path injection queries. See their `getReportingNode` predicates.

Note that the accepted test expectation changes only happen in an experimental query, where alert wobble shouldn't be a problem.

### CODE REVIEW

**Comment by atorralba** [comment]
DCA looks much better. There are some lost results that were basically repeated alerts with slightly different paths (summaries through classes from source).

File: java/ql/lib/ext/java.nio.model.yml [hunk 1]
```diff
@@ -6,6 +6,7 @@ extensions:
       - ["java.nio", "ByteBuffer", False, "array", "()", "", "Argument[this]", "ReturnValue", "taint", "manual"]
       - ["java.nio", "ByteBuffer", False, "get", "", "", "Argument[this]", "ReturnValue", "taint", "manual"]
       - ["java.nio", "ByteBuffer", False, "wrap", "(byte[])", "", "Argument[0]", "ReturnValue", "taint", "manual"]
+      - ["java.nio", "Paths", False, "get", "(URI)", "", "Argument[0]", "ReturnValue", "taint", "manual"] # old PathCreation
```
Comment (jcogs33) [inline java/ql/lib/ext/java.nio.model.yml:9]: Is this model a typo? As far as I can tell, the `Paths` class does not exist in the [`java.nio`](https://docs.oracle.com/javase/8/docs/api/java/nio/package-summary.html) package.

File: java/ql/lib/ext/java.nio.file.model.yml [hunk 2]
```diff
@@ -3,11 +3,9 @@ extensions:
       pack: codeql/java-all
       extensible: sinkModel
     data:
-      - ["java.nio.file", "Files", False, "copy", "(Path,OutputStream)", "", "Argument[0]", "path-injection", "manual"]
-      - ["java.nio.file", "Files", False, "copy", "(Path,Path,CopyOption[])", "", "Argument[0]", "path-injection", "manual"]
-      - ["java.nio.file", "Files", False, "copy", "(Path,Path,CopyOption[])", "", "Argument[1]", "path-injection", "manual"]
+      - ["java.nio.file", "Files", False, "copy", "", "", "Argument[0]", "path-injection", "manual"]
       - ["java.nio.file", "Files", False, "copy", "(InputStream,Path,CopyOption[])", "", "Argument[0]", "file-content-store", "manual"]
-      - ["java.nio.file", "Files", False, "copy", "(InputStream,Path,CopyOption[])", "", "Argument[1]", "path-injection", "manual"]
+      - ["java.nio.file", "Files", False, "copy", "", "", "Argument[1]", "path-injection", "manual"]
```
Comment (jcogs33) [inline java/ql/lib/ext/java.nio.file.model.yml:8]: Are these changes to the `java.nio.file.Files#copy` methods intentional? These seem to revert the changes in https://github.com/github/codeql/pull/13248.

**Review by jcogs33 — COMMENTED** [review]
I added a few comments/questions.

Also, it looks like the Java Language Tests are failing on the ZipSlip query's tests due to including the `File` constructor and `Path.resolve` as `path-injection` sinks.

File: java/ql/lib/ext/java.nio.model.yml (hunk 1 above)
Comment (atorralba) [reply java/ql/lib/ext/java.nio.model.yml:9]: Indeed, and this shouldn't be adding steps anyway, AND we shouldn't have a step and a sink at the same node. All in this row is a nonsense, thanks for catching 😄

**Comment by atorralba** [comment]
> Also, it looks like the Java Language Tests are failing on the ZipSlip query's tests due to including the `File` constructor and `Path.resolve` as `path-injection` sinks.

Good catch! This actually made me think for a bit, because there's no elegant solution: `File` constructors and `Path.resolve` (et al) _are_ bad sinks, but we are currently keeping them to avoid alert wobble in `java/path-injection`, while we don't want them in `java/zipslip`.

What I ended up doing is [adding an exclusion](https://github.com/github/codeql/pull/13225/commits/b75ab21ba498ee12023d9f344459094be5258a97) to `java/zipslip`, akin to the `getReportingNode` predicate in the path injection queries. It seemed like the lesser evil in this case.

File: java/ql/lib/semmle/code/java/security/ZipSlipQuery.qll [hunk 3]
```diff
@@ -40,5 +41,25 @@ module ZipSlipFlow = TaintTracking::Global<ZipSlipConfig>;
...
+  FileCreationSink() {
+    sinkNode(this, "path-injection") and
+    not isPathCreation(this)
+  }
+}
+
+/**
+ * Holds if `sink` is a path creation node that doesn't imply a read/write filesystem operation.
+ * This is to avoid creating new spurious alerts, since `PathCreation` sinks weren't
+ * previosuly part of this query.
```
Comment (jcogs33) [inline java/ql/lib/semmle/code/java/security/ZipSlipQuery.qll:53]: Tiny typo in the QLDoc
```suggestion
 * previously part of this query.
```

**Review by jcogs33 — APPROVED** [review]
LGTM now if the DCA results are good.
(Also found one small typo in a QLDoc)

File: java/ql/lib/semmle/code/java/security/ZipSlipQuery.qll (hunk 3 above)
Comment (atorralba) [reply java/ql/lib/semmle/code/java/security/ZipSlipQuery.qll]: 😭 🙇

**Comment by atorralba** [comment]
>LGTM now if the DCA results are good.

DCA actually showed an unexpected increase in ZipSlip alerts, which was caused by missing logic in `ZipSlipQuery::isPathCreation`. That's been fixed in 27763d6, and the rest of the alert diffs in the latest DCA run look reasonable (mostly, private constructors previously considered as sinks are no longer modeled because of precise signatures being used in models-as-data).
